csharp/src/Ice/Connection.cs

```diff
@@ -836,15 +833,35 @@ public class TcpConnection : IPConnection
         private SslStream? SslStream => _socket.Underlying.SslStream;
 
         internal TcpConnection(
-            IConnectionManager manager,
             Endpoint endpoint,
             MultiStreamOverSingleStreamSocket socket,
-            IConnector? connector,
-            string connectionId,
+            object? label,
             ObjectAdapter? adapter)
-            : base(manager, endpoint, socket, connector, connectionId, adapter)
+            : base(endpoint, socket, label, adapter)
         {
         }
+
+        internal override bool CanTrust(NonSecure preferNonSecure)
+        {
+            // If the connection uses ice1 endpoint, the connection is secure, or non-secure connections are always
+            // allowed we can trust the connection.
+            if (Endpoint.Protocol == Protocol.Ice1 || IsSecure || preferNonSecure == NonSecure.Always)
+            {
+                return true;
+            }
+
+            if (preferNonSecure == NonSecure.SameHost)
+            {
+                return RemoteEndpoint?.IsSameHost() ?? false;
```

Comment:
I find the names LocalEndpoint and RemoteEndpoint (not introduced in this PR) very confusing. They are IPEndPoints, not Endpoints.